include/nbl/builtin/hlsl/bxdf/base/cook_torrance_base.hlsl

```diff
@@ -0,0 +1,383 @@
+        const vector3_type localH = ndf.generateH(localV, u);
+
+        cache = anisocache_type::createForReflection(localV, localH);
+        struct reflect_wrapper
+        {
+            vector3_type operator()() NBL_CONST_MEMBER_FUNC
+            {
+                return r(VdotH);
+            }
+            bxdf::Reflect<scalar_type> r;
+            scalar_type VdotH;
+        };
+        reflect_wrapper r;
+        r.r = bxdf::Reflect<scalar_type>::create(localV, localH);
+        r.VdotH = cache.getVdotH();
+        ray_dir_info_type localL = localV_raydir.template reflect<reflect_wrapper>(r);
+
+        // fail if samples have invalid paths
+        if (localL.getDirection().z < scalar_type(0.0))  // NdotL<0
+            localL.makeInvalid(); // should check if sample direction is invalid
+
+        return sample_type::createFromTangentSpace(localL, interaction.getFromTangentSpace());
+    }
+    template<typename C=bool_constant<IsBSDF> >
+    enable_if_t<C::value && IsBSDF, sample_type> generate(NBL_CONST_REF_ARG(anisotropic_interaction_type) interaction, const vector3_type u, NBL_REF_ARG(anisocache_type) cache)
+    {
+        fresnel_type _f = impl::getOrientedFresnel<fresnel_type, IsBSDF>::__call(fresnel, interaction.getNdotV());
+        fresnel::OrientedEtaRcps<monochrome_type> rcpEta = _f.getOrientedEtaRcps();
+
+        ray_dir_info_type V = interaction.getV();
+        const vector3_type localV = interaction.getTangentSpaceV();
+        const vector3_type upperHemisphereV = ieee754::flipSignIfRHSNegative<vector3_type>(localV, hlsl::promote<vector3_type>(interaction.getNdotV()));
+        const vector3_type localH = ndf.generateH(upperHemisphereV, u.xy);
+        const vector3_type H = hlsl::mul(interaction.getFromTangentSpace(), localH);
+
+        const scalar_type VdotH = hlsl::dot(V.getDirection(), H);
+        const scalar_type reflectance = _f(hlsl::abs(VdotH))[0];
+
+        scalar_type rcpChoiceProb;
+        scalar_type z = u.z;
+        bool transmitted = math::partitionRandVariable(reflectance, z, rcpChoiceProb);
+
+        Refract<scalar_type> r = Refract<scalar_type>::create(V.getDirection(), H);
+        bxdf::ReflectRefract<scalar_type> rr;
+        rr.refract = r;
+        ray_dir_info_type L = V.reflectRefract(rr, transmitted, rcpEta.value[0]);
+
+        const vector3_type T = interaction.getT();
+        const vector3_type B = interaction.getB();
+        const vector3_type _N = interaction.getN();
+
+        // fail if samples have invalid paths
+        const vector3_type Ldir = L.getDirection();
+        const scalar_type LdotH = hlsl::dot(Ldir, H);
+        if ((ComputeMicrofacetNormal<scalar_type>::isTransmissionPath(VdotH, LdotH) != transmitted) || (LdotH * hlsl::dot(_N, Ldir) < scalar_type(0.0)))
+            L.makeInvalid(); // should check if sample direction is invalid
+        else
+            cache = anisocache_type::create(VdotH, Ldir, H, T, B, _N, transmitted);
```

Comment:
use this instead https://github.com/Devsh-Graphics-Programming/Nabla/blob/804014f28e7469036673169266a013f42e018f60/include/nbl/builtin/hlsl/bxdf/common.hlsl#L725-L740

Reply:
then you won't need all the `T,B,_N` variables